tools/hsmtools.c

```diff
@@ -0,0 +1,214 @@
+		close(fd);
+		rename("hsm_secret.backup", hsm_secret_path);
+		errx(ERROR_HSM_FILE,
+		     "Failure writing plaintext seed to hsm_secret.");
+	}
+
+	/* Be as paranoïd as in hsmd with the file state on disk. */
+	if (!ensure_hsm_secret_exists(fd)) {
+		unlink_noerr(hsm_secret_path);
+		rename("hsm_secret.backup", hsm_secret_path);
+		errx(ERROR_HSM_FILE,
+		     "Could not ensure hsm_secret existence.");
+	}
+	unlink_noerr("hsm_secret.backup");
+
+	printf("Succesfully decrypted hsm_secret, be careful now :-).\n");
+	return 0;
+}
+
+static int encrypt_hsm(const char *hsm_secret_path, const char *passwd)
+{
+	int fd;
+	struct stat st;
+	struct secret key, hsm_secret;
+	u8 salt[16] = "c-lightning\0\0\0\0\0";
+	crypto_secretstream_xchacha20poly1305_state crypto_state;
+	u8 header[crypto_secretstream_xchacha20poly1305_HEADERBYTES];
+	/* The cipher size is static with xchacha20poly1305. */
+	u8 cipher[sizeof(struct secret) + crypto_secretstream_xchacha20poly1305_ABYTES];
+
+	/* We can use a relative path as libplugin chdir into lightning_dir
+	 * before calling us. */
+	if (stat(hsm_secret_path, &st) != 0)
+		errx(errno, "Could not stat hsm_secret");
+	if (st.st_size > 32)
+		errx(errno, "hsm_secret is already encrypted");
+	fd = open(hsm_secret_path, O_RDONLY);
+	if (fd < 0)
+		errx(errno, "Could not open hsm_secret");
+	if (!read_all(fd, &hsm_secret, sizeof(hsm_secret)))
+		errx(errno, "Could not read hsm_secret");
+
+	/* Derive the encryption key from the password provided, and try to encrypt
+	 * the seed. */
+	if (crypto_pwhash(key.data, sizeof(key.data), passwd, strlen(passwd), salt,
+	                  crypto_pwhash_argon2id_OPSLIMIT_MODERATE,
+	                  crypto_pwhash_argon2id_MEMLIMIT_MODERATE,
+	                  crypto_pwhash_ALG_ARGON2ID13) != 0)
+		errx(ERROR_LIBSODIUM, "Could not derive a key from the password.");
+	if (crypto_secretstream_xchacha20poly1305_init_push(&crypto_state, header,
+	                                                    key.data) != 0)
+		errx(ERROR_LIBSODIUM, "Could not initialize the crypto state");
+	if (crypto_secretstream_xchacha20poly1305_push(&crypto_state, cipher,
+	                                               NULL, hsm_secret.data,
+	                                               sizeof(hsm_secret.data),
+	                                               NULL, 0, 0) != 0)
+		errx(ERROR_LIBSODIUM, "Could not encrypt the seed.");
+
+	/* Create a backup file, "just in case". */
+	rename(hsm_secret_path, "hsm_secret.backup");
+	fd = open(hsm_secret_path, O_CREAT|O_EXCL|O_WRONLY, 0400);
+	if (fd < 0)
+		errx(ERROR_HSM_FILE, "Could not open new hsm_secret");
+
+	/* Write the encrypted hsm_secret. */
+	if (!write_all(fd, header, sizeof(header))
+		|| !write_all(fd, cipher, sizeof(cipher))) {
+		unlink_noerr(hsm_secret_path);
+		close(fd);
+		rename("hsm_secret.backup", hsm_secret_path);
+		errx(ERROR_HSM_FILE, "Failure writing cipher to hsm_secret.");
+	}
+	if (!write_all(fd, cipher, sizeof(cipher))) {
```

Comment:
i don't think you need this, you're writing the cipher in the previous if statement as well.

Reply:
I need to write both the header and the body of the cipher : the variable name is maybe not very explicit..

Reply:
yeah you write the cipher already in L161

Reply:
Oh yes, I forgot to remove it here thanks